cmd/mirroring/main.go

```diff
@@ -79,85 +119,104 @@ func main() {
-		// File containing old snapshot exists
-		if err := readLogInfo(&treeSize, &root); err != nil {
+		// File containing previous checkpoints exists
+		sth, err = readLatestCheckpoint(*logInfoFile)
+		if err != nil {
 			log.Fatalf("Reading log info: %v", err)
 		}
 	} else if errors.Is(err, fs.ErrNotExist) {
-		// No old snapshot data available: get latest signed tree head and load
+		// No old snapshot data available, get latest checkpoint
 		logInfo, err := mirroring.GetLogInfo(rekorClient)
 		if err != nil {
 			log.Fatalf("Getting log info: %v", err)
 		}
-
-		pubkey, err := mirroring.GetPublicKey(rekorClient)
-		if err != nil {
-			log.Fatalf("Getting public key: %v", err)
-		}
-
-		// Verify the queried signed tree head with server's public key
-		if err := mirroring.VerifySignedTreeHead(logInfo, pubkey); err != nil {
-			log.Fatalf("Verifying signed tree head: %v", err)
+		sth = &util.SignedCheckpoint{}
+		if err := sth.UnmarshalText([]byte(*logInfo.SignedTreeHead)); err != nil {
+			log.Fatalf("unmarshalling logInfo.SignedTreeHead to Checkpoint: %v", err)
 		}
-
-		treeSize = *logInfo.TreeSize
-		root = *logInfo.RootHash
 		first = true
 	} else {
 		// Any other errors reading the file
-		log.Fatalf("Reading %q: %v", *logInfoFile, err)
+		log.Fatalf("reading %q: %v", *logInfoFile, err)
 	}
 
-	// Open file to create/append new snapshots
-	file, err := os.OpenFile(*logInfoFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	pubkey, err := mirroring.GetPublicKey(rekorClient)
 	if err != nil {
-		log.Println(err)
+		log.Fatalf("getting public key: %v", err)
 	}
-	defer file.Close()
+
+	// TODO: Verify using public key from TUF
+	// Verify the checkpoint with the server's public key
+	if err := mirroring.VerifySignedTreeHead(sth, pubkey); err != nil {
+		log.Fatalf("verifying checkpoint: %v", err)
+	}
+	log.Printf("Current checkpoint verified - Tree Size: %d Root Hash: %s\n", sth.Size, hex.EncodeToString(sth.Hash))
 
 	// If this is the very first snapshot within the monitor, save the snapshot
 	if first {
-		if _, err := file.WriteString(fmt.Sprintf("%d %s\n", treeSize, root)); err != nil {
-			log.Fatalf("Failed to write to file: %v", err)
+		s, err := sth.SignedNote.MarshalText()
+		if err != nil {
+			log.Fatalf("failed to marshal checkpoint: %v", err)
+		}
+
+		// Open file to create new snapshot
+		file, err := os.OpenFile(*logInfoFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			log.Fatalf("failed to open log file: %v", err)
+		}
+		defer file.Close()
+
+		// Replace newlines to flatten checkpoint to single line
+		if _, err := file.WriteString(fmt.Sprintf("%s\n", strings.Replace(string(s), "\n", "\\n", -1))); err != nil {
+			log.Fatalf("failed to write to file: %v", err)
 		}
 	}
 
 	for {
 		// Check for root hash consistency
-		newTreeSize, newRoot, err := mirroring.VerifyLogConsistency(rekorClient, treeSize, root)
+		newSTH, err := mirroring.VerifyLogConsistency(rekorClient, int64(sth.Size), sth.Hash)
```

Comment:
meta comment: the overall pattern in the first STH case is very strange: get an STH, then verifylogconsistency with another newSTH retrieved seconds later?

if this is the first, then why not JUST get an STH, verify it, write it, and exit. if there's existing, load the old state, and verify consistency with a new one and write.

so something like:
1. Get new STH & verify
2. if old STH if exists, load and verify log consistency
3. Write new STH

This avoid the two `GetLogInfo`s that happen on first case.

Reply:
Yep, very much agree! I was getting complacent with the existing code. Lemme add a TODO to clean that up in the next PR to avoid this PR being too messy.